sml/src/sml_list.c

```diff
@@ -112,16 +112,64 @@ sml_list *sml_list_init() {
 
 void sml_list_add(sml_list *list, sml_list *new_entry) { list->next = new_entry; }
 
+/* Note: this gets passed _every_ object found */
+static bool detect_buggy_dzg_dvs74(sml_list *l)
+{
+#define DZG_SERIAL_LENGTH	10
+	static const unsigned char dzg_serial_oid[] = {1, 0, 96, 1, 0, 255};
+	static const unsigned char dzg_serial_start[] = { 0x0a, 0x01, 'D', 'Z', 'G' };
+	static const struct {
+		uint64_t start, end;
+	} broken_dvs74_ranges[] = {
+		{ .start = 42000000, .end = 48999999 },
+		{ .start = 55000000, .end = 58999999 },
+	};
+#define NUM_BROKEN_RANGES (sizeof(broken_dvs74_ranges)/sizeof(broken_dvs74_ranges[0]))
+	const unsigned char *val;
+	uint64_t serial = 0;
+	unsigned int i;
+
+	/* check that object ID is for the serial number */
+	if (!l->obj_name || l->obj_name->len != sizeof(dzg_serial_oid) ||
+	    memcmp(l->obj_name->str, dzg_serial_oid, sizeof(dzg_serial_oid)))
+		return false;
+
+	/* check that it's an octet string of the right length */
+	if (!l->value || l->value->type != SML_TYPE_OCTET_STRING ||
+	    l->value->data.bytes->len != DZG_SERIAL_LENGTH)
+		return false;
+
+	val = l->value->data.bytes->str;
+
+	/* check that it's a DZG meter at all */
+	if (memcmp(val, dzg_serial_start, sizeof(dzg_serial_start)))
+		return false;
+
+	/* convert remainder of serial number to integer */
+	for (i = sizeof(dzg_serial_start); i < DZG_SERIAL_LENGTH; i++) {
+		serial <<= 8;
+		serial |= val[i];
+	}
+
+	/*
+	 * Check that it's in one of the affected ranges,
+	 * which means it's a DVS74 with the bug
+	 * (DVS74 with serial >= 60000000 have the bug fixed)
+	 */
+	for (i = 0; i < NUM_BROKEN_RANGES; i++) {
+		if (serial >= broken_dvs74_ranges[i].start &&
+		    serial <= broken_dvs74_ranges[i].end)
+			return true;
+	}
+
+	return false;
+}
+
 struct workarounds {
-	unsigned int old_dzg_meter : 1;
+	unsigned int old_dzg_dvs74 : 1;
```

Comment:
i got confused vs. https://github.com/volkszaehler/libsml/pull/133 , i had not realized that the "workarounds" struct already existed before #133 ) :astonished: